<a href="https://colab.research.google.com/github/sergiocostaifes/PPCOMP_DM/blob/main/notebooks/05_feature_engineering.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 05_feature_engineering.ipynb — Engenharia de Atributos (5 min)

## Objetivo
Construir atributos derivados para modelagem supervisionada a partir da **série temporal contínua (5 minutos)** e da segmentação de
episódios críticos, adotando **taxa de falha** como métrica base (**fail_rate**), garantindo consistência com a definição de episódio do Notebook 04.

## Entradas (artefatos dos Notebooks 03 e 04)
- `window_5min_series.parquet`
- `episodes_detected.parquet`

## Saídas (artefatos deste Notebook)
- `window_5min_features.parquet`
- `05_feature_engineering_summary.json`

## Compatibilidade com o pipeline existente (00 → 04)
Este notebook mantém a granularidade oficial de 5 minutos e:
- Preserva `bucket_id` como eixo temporal
- Reproduz a variável-alvo `is_critical` a partir de `episodes_detected.parquet` (sem redefinir critério)
- Define **`fail_rate = n_failed / n_events`** como métrica base para engenharia de atributos
- Gera atributos temporais (lags, rolling), variação percentual e normalização global sobre `fail_rate`
- Mantém `n_failed` e `n_events` como atributos brutos (não derivados) para auditoria e diagnósticos

## Definições
- Janela temporal: 5 min (300 s)
- Métrica base para features: `fail_rate = n_failed / n_events`
  - Tratamento de borda: quando `n_events = 0`, define-se `fail_rate = 0.0`
- Variável-alvo (`is_critical`): 1 se `bucket_id` estiver contido em algum intervalo `[start_bucket, end_bucket]` de `episodes_detected.parquet`, caso contrário 0
- Rolling 1h: 12 janelas de 5 minutos

## Features construídas (sobre fail_rate)
1. Lags temporais (1, 2 e 3 janelas)
2. Rolling mean (1h)
3. Rolling std (1h)
4. Variação percentual
5. Z-score global (normalização por μ e σ calculados sobre `fail_rate`)

## Observações de reprodutibilidade
- Bootstrap via Drive + repositório `PPCOMP_DM`
- Uso de `src.paths` e `ensure_dirs()`
- Persistência em `03-features`
- Resumo estatístico salvo em `04-reports`
- Seed fixa

In [3]:
# ============================================================
# 05_feature_engineering.ipynb
# Engenharia de atributos (5-min)
# Métrica base: fail_rate = n_failed / n_events
# Pipeline PPCOMP_DM (Google Cluster Trace)
# ============================================================

# -----------------------------
# 0) BOOTSTRAP (Colab + Repo)
# -----------------------------

from pathlib import Path
import os
import sys
import subprocess
import importlib
import random
import numpy as np
import pandas as pd
import json

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# Mount Google Drive
if not Path("/content/drive/MyDrive").exists():
    from google.colab import drive
    drive.mount("/content/drive")
else:
    print("[Bootstrap] Google Drive já montado.")

REPO_DIR = Path("/content/drive/MyDrive/Mestrado/PPCOMP_DM")
GITHUB_REPO = "https://github.com/sergiocostaifes/PPCOMP_DM.git"

if not REPO_DIR.exists():
    REPO_DIR.parent.mkdir(parents=True, exist_ok=True)
    print(f"[Bootstrap] Clonando repositório em: {REPO_DIR}")
    subprocess.run(["git", "clone", GITHUB_REPO, str(REPO_DIR)], check=True)
else:
    try:
        print("[Bootstrap] Atualizando repositório (git pull)...")
        subprocess.run(["git", "-C", str(REPO_DIR), "pull"], check=True)
    except Exception as e:
        print("[Bootstrap] Aviso: não foi possível atualizar:", e)

os.chdir(str(REPO_DIR))
print("[Bootstrap] CWD =", os.getcwd())

repo_str = str(REPO_DIR)
if repo_str not in sys.path:
    sys.path.insert(0, repo_str)

importlib.invalidate_caches()

from src.paths import FEATURES_PATH, REPORTS_PATH, ensure_dirs
ensure_dirs()

print("FEATURES_PATH =", FEATURES_PATH)
print("REPORTS_PATH  =", REPORTS_PATH)

def log(msg: str) -> None:
    print(f"[05_feature_engineering] {msg}")

# =========================
# 1) Leitura da série base + episódios
# =========================

SERIES_FILE = FEATURES_PATH / "window_5min_series.parquet"
EPISODES_FILE = FEATURES_PATH / "episodes_detected.parquet"

assert SERIES_FILE.exists(), f"Arquivo não encontrado: {SERIES_FILE}"
assert EPISODES_FILE.exists(), f"Arquivo não encontrado: {EPISODES_FILE}"

df = pd.read_parquet(SERIES_FILE)
episodes = pd.read_parquet(EPISODES_FILE)

df = df.sort_values("bucket_id").reset_index(drop=True)

# Colunas mínimas
assert "bucket_id" in df.columns
assert "n_failed" in df.columns
assert "n_events" in df.columns

log(f"Shape série base: {df.shape}")

# =========================
# 2) Definir fail_rate (métrica base coerente)
# =========================
den = df["n_events"].replace(0, np.nan)
df["fail_rate"] = (df["n_failed"] / den).fillna(0.0).astype("float32")

# =========================
# 3) Criar variável target is_critical a partir dos episódios
# =========================
df["is_critical"] = 0

# Marca 1 dentro de [start_bucket, end_bucket]
# (mantém a mesma lógica do NB05 atual, apenas preservando coerência com NB04)
for _, row in episodes.iterrows():
    df.loc[
        (df["bucket_id"] >= row["start_bucket"]) &
        (df["bucket_id"] <= row["end_bucket"]),
        "is_critical"
    ] = 1

log(f"Total janelas críticas: {int(df['is_critical'].sum())}")

# =========================
# 4) Construção de features (SOMENTE sobre fail_rate)
# =========================
BASE = "fail_rate"

# Lags
df["lag_1"] = df[BASE].shift(1)
df["lag_2"] = df[BASE].shift(2)
df["lag_3"] = df[BASE].shift(3)

# Rolling 1 hora (12 janelas de 5 min)
W = 12
df["rolling_mean_1h"] = df[BASE].rolling(W, min_periods=1).mean()
df["rolling_std_1h"] = df[BASE].rolling(W, min_periods=1).std().fillna(0.0)

# Variação percentual
df["pct_change"] = df[BASE].pct_change()
df["pct_change"] = df["pct_change"].replace([np.inf, -np.inf], 0.0).fillna(0.0)

# Normalização global (sobre fail_rate)
mu = df[BASE].mean()
sigma = df[BASE].std(ddof=0)
df["zscore_global"] = (df[BASE] - mu) / (sigma if sigma > 0 else 1.0)

# =========================
# 5) Limpeza inicial (remover NaNs de lag)
# =========================

df = df.dropna().reset_index(drop=True)

log(f"Shape após feature engineering: {df.shape}")

# =========================
# 6) Persistência
# =========================

OUT_FILE = FEATURES_PATH / "window_5min_features.parquet"
df.to_parquet(OUT_FILE, compression="snappy", index=False)

summary = {
    "rows": int(len(df)),
    "features": list(df.columns),
    "critical_ratio": float(df["is_critical"].mean()),
    "base_metric": "fail_rate",
    "mu": float(mu),
    "sigma": float(sigma),
}

summary_file = REPORTS_PATH / "05_feature_engineering_summary.json"
summary_file.write_text(json.dumps(summary, indent=2, ensure_ascii=False))

log("Notebook 05 finalizado com sucesso.")
df.head()

[Bootstrap] Google Drive já montado.
[Bootstrap] Atualizando repositório (git pull)...
[Bootstrap] CWD = /content/drive/MyDrive/Mestrado/PPCOMP_DM
FEATURES_PATH = /content/drive/MyDrive/Mestrado/02-datasets/03-features
REPORTS_PATH  = /content/drive/MyDrive/Mestrado/04-reports
[05_feature_engineering] Shape série base: (8918, 18)
[05_feature_engineering] Total janelas críticas: 394
[05_feature_engineering] Shape após feature engineering: (8914, 27)
[05_feature_engineering] Notebook 05 finalizado com sucesso.


,bucket_id,bucket_start_us,n_events,n_failed,n_machines,n_collections,mean_priority,mean_req_cpus,mean_req_mem,req_cpus_presence_rate,...,event_KILL_count,fail_rate,is_critical,lag_1,lag_2,lag_3,rolling_mean_1h,rolling_std_1h,pct_change,zscore_global
0,15,4500000000,44,10,41,29,249.522727,0.008850,0.003220,1.0,...,0,0.227273,0,0.323529,0.282609,0.240000,0.268353,0.043738,-0.297521,0.218397
1,16,4800000000,22,5,22,19,219.045455,0.011749,0.003593,1.0,...,1,0.227273,0,0.227273,0.323529,0.282609,0.260137,0.042099,0.000000,0.218397
2,17,5100000000,28,5,27,22,279.392857,0.010309,0.004613,1.0,...,0,0.178571,0,0.227273,0.227273,0.323529,0.246543,0.050266,-0.214286,-0.127306
3,18,5400000000,29,7,29,22,259.517241,0.007249,0.009697,1.0,...,2,0.241379,0,0.178571,0.227273,0.227273,0.245805,0.045928,0.351724,0.318531
4,19,5700000000,36,9,34,27,237.027778,0.010176,0.009284,1.0,...,0,0.250000,0,0.241379,0.178571,0.227273,0.246329,0.042547,0.035714,0.379725


## Achados do Notebook 05 — Engenharia de Atributos (base: taxa)

### Entrada e consistência temporal
- Série temporal contínua com **8918** janelas de 5 minutos.
- Após construção de lags e remoção das janelas iniciais indefinidas: **8914** registros finais.
- Total de atributos finais: **27 colunas**.
- A estrutura temporal foi preservada sem embaralhamento ou vazamento de informação futura.

### Métrica base
- Métrica oficial para engenharia de atributos: **fail_rate = n_failed / n_events**
- Tratamento de borda: fail_rate = 0 quando n_events = 0
- μ global (fail_rate): **0.1965058446**
- σ global (fail_rate): **0.1408760697**

### Distribuição da variável-alvo
- Total de janelas críticas: **394**
- Proporção de janelas críticas: **4,42%**

Observação: trata-se de um problema naturalmente desbalanceado, coerente com a detecção de eventos raros em sistemas distribuídos.

### Features construídas (sobre fail_rate)

1. Lags temporais (1, 2 e 3 janelas)
2. Rolling mean (1 hora = 12 janelas)
3. Rolling std (1 hora)
4. Variação percentual
5. Z-score global

Essas features capturam:

- Inércia temporal
- Tendência local
- Volatilidade de curto prazo
- Intensidade relativa ao baseline global

### Sanidade estatística

O valor médio da taxa (~0,1965) indica que, em média, cerca de 19,6% dos eventos por janela resultam em falha.
O desvio padrão (~0,1409) demonstra variabilidade significativa entre janelas, compatível com a presença de picos episódicos detectados no NB04.

O rolling_std_1h permite capturar mudanças locais de regime operacional.
O zscore_global fornece medida padronizada da intensidade relativa ao comportamento médio da série.

### Considerações metodológicas

A engenharia de atributos mantém:

- Causalidade temporal
- Reprodutibilidade
- Granularidade fixa de 5 minutos
- Coerência total com a definição de episódio baseada em taxa

O dataset `window_5min_features.parquet` constitui a base para:

- Rotulagem temporal (BEFORE / DURING / AFTER)
- Treinamento supervisionado
- Avaliação comparativa de classificadores

### Conclusão

O Notebook 05 consolida a transição entre detecção estatística baseada em taxa (NB04) e modelagem preditiva supervisionada, garantindo alinhamento conceitual entre definição de criticidade e construção dos atributos explicativos.